server/task.go

```diff
@@ -718,8 +718,16 @@ func (s *Server) patchTaskStatus(ctx context.Context, task *api.Task, taskStatus
 		if pipeline == nil {
 			return nil, errors.Errorf("pipeline not found for ID %v", taskPatched.PipelineID)
 		}
-		lastStage := pipeline.StageList[len(pipeline.StageList)-1]
-		if lastStage.TaskList[len(lastStage.TaskList)-1].ID == taskPatched.ID {
+		taskAllDone := true
```

Comment:
Move this to a separate function and have a unit test. This can reduce the length of the function and have some unit test coverage.